test/jdk/java/lang/ProcessBuilder/PipelineLeaksFD.java

```diff
@@ -0,0 +1,135 @@
+                        new ProcessBuilder("cat"),
+                        new ProcessBuilder("cat"),
+                        new ProcessBuilder("cat"),
+                        new ProcessBuilder("cat"))},
+        };
+    }
+
+    @Test(dataProvider = "builders")
+    void checkForLeaks(List<ProcessBuilder> builders) throws IOException {
+
+        Set<PipeRecord> pipesBefore = myPipes();
+        if (pipesBefore.size() < 3) {
+            System.out.println(pipesBefore);
+            Assert.fail("There should be at least 3 pipes before, (0, 1, 2)");
+        }
+
+        List<Process> processes = ProcessBuilder.startPipeline(builders);
+
+        OutputStream out = processes.get(0).getOutputStream();
+        out.write('a');
+        out.close();
+
+        Process last = processes.get(processes.size() - 1);
+        try (InputStream inputStream = last.getInputStream()) {
+            byte[] bytes = inputStream.readAllBytes();
+            Assert.assertEquals(bytes.length, 1, "Bytes read");
+        }
+
+        processes.forEach(p -> waitForQuiet(p));
+
+        Set<PipeRecord> pipesAfter = myPipes();
+        if (!pipesBefore.equals(pipesAfter)) {
+            Set<PipeRecord> missing = new HashSet<>(pipesBefore);
+            missing.removeAll(pipesAfter);
+            printPipes(missing, "Missing from pipesAfter");
+            Set<PipeRecord> extra = new HashSet<>(pipesAfter);
+            extra.removeAll(pipesBefore);
+            printPipes(extra, "Extra pipes in pipesAfter");
+            Assert.fail("More or fewer pipes than expected");
+        }
+    }
+
+    static void printPipes(Set<PipeRecord> pipes, String label) {
+        System.out.printf("%s: [%d]%n", label, pipes.size());
+        pipes.forEach(r -> System.out.printf("%20s: %20s%n", r.fd(), r.link()));
+    }
+
+    static void waitForQuiet(Process p) {
+        try {
+            int st = p.waitFor();
+            if (st != 0) {
+                System.out.println("non-zero exit status: " + p);
+            }
+        } catch (InterruptedException ie) {
+        }
+    }
+
+    /**
+     * Collect a Set of pairs of /proc fd paths and the symbol links that are pipes.
+     * @return A set of PipeRecords, possibly empty
+     * @throws IOException if reading the directory entries of "/proc/<pid>/fd/*" fails
+     */
+    static Set<PipeRecord> myPipes() throws IOException {
+        Path path = Path.of("/proc/" + ProcessHandle.current().pid() + "/fd");
+        Stream<Path> s = Files.walk(path);
+        return s.filter(Files::isSymbolicLink)
+                .map(p -> {
+                    try {
+                        return new PipeRecord(p, Files.readSymbolicLink(p));
+                    } catch (IOException ioe) {
+                    }
+                    return new PipeRecord(p, null);
+                })
+                .filter(p1 -> p1.link().toString().startsWith("pipe:"))
+                .collect(Collectors.toSet());
```

Comment:
Is this intentionally leaking the returned `DirectoryStream` from `Files.walk` to  trigger the previous failure mode or should this be auto-closed (i.e. https://errorprone.info/bugpattern/StreamResourceLeak )?

```suggestion
        try (Stream<Path> s = Files.walk(path)) {
            return s.filter(Files::isSymbolicLink)
                    .map(p -> {
                        try {
                            return new PipeRecord(p, Files.readSymbolicLink(p));
                        } catch (IOException ioe) {
                        }
                        return new PipeRecord(p, null);
                    })
                    .filter(p1 -> p1.link().toString().startsWith("pipe:"))
                    .collect(Collectors.toSet());
        }
```

Reply:
@schlosna An oversight,  thanks for the reminder.